File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        row = [{} for _ in range(9)]
        col = [{} for _ in range(9)]
        grid_3x3 = [{} for _ in range(9)]
        
        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num == '.':
                    continue
                
                k = (i // 3) * 3 + (j // 3)
                if num in row[i] or num in col[j] or num in grid_3x3[k]:
                    return False
                
                row[i][num] = col[j][num] = grid_3x3[k][num] = 1
        
        return True

File: LeetCode/0036-valid-sudoku/0036-valid-sudoku.java
public class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] row = new int[9][9];
        int[][] col = new int[9][9];
        int[][] grid_3x3 = new int[9][9];
        
        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                if(board[i][j] == '.') {
                    continue;
                }
                int num = board[i][j] - '1';
                int k = (i / 3) * 3 + (j / 3);
                if(row[i][num] == 1 || col[j][num] == 1 || grid_3x3[k][num] == 1) {
                    return false;
                }
                row[i][num] = col[j][num] = grid_3x3[k][num] = 1;
            }
        }
        return true;
    }
}